libs/storage/include/storage/have_seen_recently_cache.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <deque>
+#include <unordered_set>
+
+namespace fetch {
+namespace storage {
+
+/**
+ * Not thread safe.
+ *
+ * This class provides the functionality that you can add elements to it,
+ * they will be held in the order they were added up to a fixed size,
+ * and you can query whether they are still there (have been recently seen).
+ *
+ * Due to the implementation behaviour, adding the same item when it is
+ * already Seen may result in it returning it as not been seen sooner than expected.
+ *
+ */
+template <typename T>
+class HaveSeenRecentlyCache
```

Comment:
Is `Have` really needed in the name? Smth like `RecentlySeenCache` would be easier to pronounce imo.

Reply:
Makes sense. Dropped the 'have' in the class name.